File: algorithms/javaalgo/src/test/java/programmers/level2/numberofsumssubsequences/TheNumberOfSumsOfSubsequencesTest.java
package programmers.level2.numberofsumssubsequences;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.HashSet;
import java.util.Set;
import org.junit.jupiter.api.Test;

public class TheNumberOfSumsOfSubsequencesTest {

    @Test
    void test() {
        int[] elements = {7, 9, 1, 1, 4};

        assertThat(solution(elements)).isEqualTo(18);
    }

    public int solution(int[] elements) {
        Set<Integer> answers = new HashSet<>();
        int length = elements.length;
        for (int size = 1; size <= length; size++) {
            for (int i = 0; i < length; i++) {
                int sum = 0;
                for (int j = i; j < i + size; j++) {
                    sum += elements[j % length];
                }
                answers.add(sum);
            }
        }

        return answers.size();
    }
}

File: algorithms/javascriptalgo/programmers/level2/theNumberOfSumsOfSubsequences.test.js
test("test", () => {
    let elements = [7, 9, 1, 1, 4];

    expect(solution(elements)).toBe(18);
});

function solution(elements) {
    let answer = new Set();
    let length = elements.length;
    for (let size = 1; size <= length; size++) {
        for (let i = 0; i < length; i++) {
            let sum = 0;
            for (let j = i; j < i + size; j++) {
                sum += elements[j % length];
            }
            answer.add(sum);
        }
    }

    return answer.size;
}
